### ISSUE TITLE
Publish to F-Droid

### ISSUE THREAD

**Issue by shuvashish76** [iamr0s/Dhizuku#11]
您会考虑将此应用发布到 F-Droid 吗？

**Comment by iamr0s** [comment]
可能

**Comment by shuvashish76** [comment]
@iamr0s Could you add fastlane metadata as mentioned [here](https://gitlab.com/-/snippets/1895688)?

例如，您可以查看它们的结构。https://github.com/JunkFood02/Seal/tree/main/fastlane/metadata/android

目前的描述对用户来说还不清楚。

**Comment by shuvashish76** [comment]
@iamr0s 任何更新？

**Comment by theimpulson** [comment]
Bumping this up, @iamr0s is this still planned?

**Comment by s1204IT** [comment]
https://gitlab.com/fdroid/rfp/-/issues/2742

We can check our progress here.

**Comment by shuvashish76** [comment]
@iamr0s From IzzySoft scanner [report](https://gitlab.com/IzzyOnDroid/repo/-/issues/382#note_1900653329)

```
SigningBlock blobs:
-------------------
0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)
```

Would be great if you could get rid of that:

```
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains.

Also need explanation regarding the permissions

```
Permissions:
------------
* android.permission.POST_NOTIFICATIONS
* android.permission.FOREGROUND_SERVICE
* android.permission.RECEIVE_BOOT_COMPLETED
* android.permission.READ_EXTERNAL_STORAGE
* android.permission.WRITE_EXTERNAL_STORAGE
* android.permission.MANAGE_EXTERNAL_STORAGE
* android.permission.QUERY_ALL_PACKAGES
* com.rosan.dhizuku.DYNAMIC_RECEIVER_NOT_EXPORTED_PERMISSION
* moe.shizuku.manager.permission.API_V23
```
---
BTW live on IzzyOnDroid: https://apt.izzysoft.de/fdroid/index/apk/com.rosan.dhizuku

**Comment by IzzySoft** [comment]
There are also some issues with the current fastlane. For example, the icon is a 0 byte file, and the description is a bit incomplete. For `en-US/full_description.txt`, may I recommend something like this:

```html
<p><i>Dhizuku</i> is a tool that allows you to share device owner permissions with other apps.</p><p>This app is similar to Shizuku (ADB) and Magisk (Root) but instead, it shares device owner permissions with other apps. An example use would be having a new Installer in the Aurora Store for allowing devices such as those sold in Japan (and those where normal package installation is blocked) to install apps from third-party app stores.</p><p><br><b>Enable service</b></p><p>You can use the following methods to enable the service:</p><ul><li>Shizuku</li><li>ADB</li></ul>
```

or, if you prefer "plain text" rendering properly (your current one will have additional line-breaks in the bullet-point list, as F-Droid replaces all newlines with `<br>` – which is why above HTML is kept on a single line):

```markdown
<i>Dhizuku</i> is a tool that allows you to share device owner permissions with other apps.

This app is similar to Shizuku (ADB) and Magisk (Root) but instead, it shares device owner permissions with other apps. An example use would be having a new Installer in the Aurora Store for allowing devices such as those sold in Japan (and those where normal package installation is blocked) to install apps from third-party app stores.

<b>Enable service</b>

You can use the following methods to enable the service:

* Shizuku
* ADB
```

That's the description currently used at IzzyOnDroid. Pulling from fastlane is enabled there, but icon and full description are currently ignored because of the above.

**Comment by IzzySoft** [comment]
Scanner just yielded again:

```
! repo/com.rosan.dhizuku_13.apk declares sensitive permission(s):
  android.permission.READ_EXTERNAL_STORAGE android.permission.MANAGE_EXTERNAL_STORAGE
```

could you please clarify what those permissions are needed for? Thanks in advance!

**Comment by IzzySoft** [comment]
And again. So shall we just led the warnings show – or could you provide a simple description on what they're needed/used for?

![Image](https://github.com/user-attachments/assets/748c7ec1-7673-4abc-a7a3-9466263676be)

**Comment by s1204IT** [comment]
Should I add an explanation for each permission?

**Comment by IzzySoft** [comment]
That would be ideal – we'd then add it for each one here (as it's already done for `QUERY_ALL_PACKAGES`, as you can see in the screenshot). But we mostly care for the "more sensitive ones" – in your case, the two chocolate colored in above screenshot. It's about transparency: less "suspicion" when things are made transparent. That's one of our "core concerns" at IzzyOnDroid 😉 Thanks a lot!

**Comment by s1204IT** [comment]
Okay, maybe not right now, but I'll get it ready eventually.

**Comment by IzzySoft** [comment]
Thanks!

**Comment by s1204IT** [comment]
Should I add that to my privacy policy?

**Comment by IzzySoft** [comment]
Great idea, thanks! I can copy over from there then. Saw several project putting it there indeed. I've just looked at the Readme and the top-level files again (did that before) and could not see the privacy policy, but your mentioning it meant it must be somewhere (thanks for mentioning it! Found it in the `docs/` folder now). Might be a good idea to link there from the Readme, too. It shows that you indeed care (many projects don't care to set up one – and I need to point a finger to my own face there, too, as I guess I forgot it with several of my own projects as well: being a privacy proponent, it's obviously too natural for me how to act privacy-wise that I didn't think of writing one 🙈 I'll definitely need to check that on my end, thanks for the pointer!)

**Comment by s1204IT** [comment]
Almost all of the permissions are used for relaying other apps, such as installing APKs, resetting the screen lock password, and running the Dhizuku service at all times during startup and execution.

**Comment by s1204IT** [comment]
Added https://github.com/iamr0s/Dhizuku/commit/f3a351b

**Comment by IzzySoft** [comment]
Thanks! Updating here now. Small question of understanding:

> Required to get the DPCs list.

What does DPC stand for here?

> - `READ_EXTERNAL_STORAGE`
> - `WRITE_EXTERNAL_STORAGE`
> - `MANAGE_EXTERNAL_STORAGE`
>
> Required to retrieve the complete file data.

Why are WRITE and MANAGE needed to **retrieve** data? And what "file data" is meant here? The other parts were added here now and are already live:

![Image](https://github.com/user-attachments/assets/e0fab0e3-ee4b-4acf-916a-b46b5f34d7aa)

Another question: is there a dedicated icon available for the app listing?

**Comment by s1204IT** [comment]
> What does DPC stand for here?

**DPC** is **D**evice **P**olicy **C**ontroller. (ex: TestDPC)

> Why are WRITE and MANAGE needed to retrieve data? And what "file data" is meant here? The other parts were added here now and are already live:

This needs to be re-investigated. Please wait a little longer.

> Another question: is there a dedicated icon available for the app listing?

Does this actually refer to the icon of Dhizuku itself?

**Comment by s1204IT** [comment]
Regarding storage permissions, these are not used by Dhizuku itself, but are used by external apps that use Dhizuku's API (such as TestDPC/InstallerX).

Therefore, the core of Dhizuku itself needs permissions. The same reason applies to other permissions.

**Comment by IzzySoft** [comment]
> DPC is Device Policy Controller. (ex: TestDPC)

Ah! I *knew* it rang a bell, but I could not locate where 🙈 Thanks!

> This needs to be re-investigated. Please wait a little longer.

Sure, will do –and thanks for digging in!

> Does this actually refer to the icon of Dhizuku itself?

Take a look [here](https://apt.izzysoft.de/packages/com.rosan.dhizuku):

![Image](https://github.com/user-attachments/assets/1ffaf789-4da0-41a0-90f6-9905dc22c6d1)

The empty space in the upper left corner (next to the first 2 paragraphs of the app description) usually shows an app icon. Same in the lists. So what I mean is not "on device", but rather "for presentation".

> Regarding storage permissions, these are not used by Dhizuku itself, but are used by external apps that use Dhizuku's API

which in turn would mean: an app that does not request any storage permission by itself could "gain storage access" through Dhizuku? For what specific purpose (so we can clarify that – as without clarification, it sounds a bit "scary")?

**Comment by IzzySoft** [comment]
> Ah, it's banner?

no, an icon. For reference:

![Image](https://github.com/user-attachments/assets/79a94ab0-a69a-4ec3-89ad-87d2dd468d67)
![Image](https://github.com/user-attachments/assets/81bb4851-1dd7-4a6d-a738-9d693b520319)

For the permissions: so it's read/write storage access _triggered_ by the "calling app" and performed by the DPC – but the "calling app" itself cannot use it to e.g. write/modify "arbitrary data" on the device? OTOH:

> At the very least, you need to declare the use of Dhizuku.

The person owning the device needs to explicitly permit that "calling app" to act as "device owner", right. And an "owner" is supposed to have "full access" (aka MANAGE). 🤔 So the intro in the privacy policy should be understand as:

> Basically, the following permissions are used to support client applications that use Dhizuku, rather than for Dhizuku itself. **They are passed to the client application, if you grant it the device ownership.**

and I'd add "passed to the client application" on our side, maybe? Would that fit?

**Comment by s1204IT** [comment]
Yes, it should be explicitly stated that these permissions are not used by Dhizuku itself.

**Comment by IzzySoft** [comment]
The app description even says it, in a way: "it shares device owner permissions with other apps". Once digging into it, it becomes clear(er). Not obvious to the "newcomer", though. Maybe the app description can be expanded a bit (I've added the last remaining "chocolate permission" to the "green list" now, so at least there are no more "warnings").

Maybe we change the second paragraph of the app description to

> This app is similar to Shizuku (ADB) and Magisk (Root) but instead, it shares device owner permissions with other apps **– making most permissions it requests available to them.** An example use would be having a new Installer in the Aurora Store for allowing devices where normal package installation is blocked to install apps from third-party app stores.

(bold text inserted by me – maybe it should even be bold in the description, to catch it easier?) I wrote "most", as I'm not sure if it would really be "all" – my guess is some of them are excluded (e.g. foreground service and Shizuku are not "given over").

Note that in my suggested rewrite, I also removed the "such as those sold in Japan" part, to make the sentence easier to follow. I'm of course open to further improvements there. And if you like, I can then offer you a PR shipping the resulting metadata in Fastlane structure, so they'd be in your hands to define how the app is presented (as you just renamed the issue to "Publish to F-Droid": F-Droid would ask for exactly this. And before you get confused: I'm calling in from IzzyOnDroid, the largest 3rd party repo, not from F-Droid.org itself 😉).

**Comment by s1204IT** [comment]
I understand. Thank you for your suggestion. As for fastlane, I will rebuild it when the icon is completed.

**Comment by IzzySoft** [comment]
Great, thanks! So you don't need my PR with a Fastlane starter package then (providing what has been set up at IzzyOnDroid currently)?

**Comment by s1204IT** [comment]
Well, at least until the icon is made.

**Comment by IzzySoft** [comment]
OK, some good news for you: I've just managed to establish [Reproducible Builds](https://reproducible-builds.org/) for Dhizuku (for some background, see e.g.  [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Here's the recipe I used:

```yaml
build:
  - sed -r '/signingConfigs.getByName/d' -i app/build.gradle.kts
  - ./gradlew assembleRelease
  - git clone -b v0.3.0 https://github.com/obfusk/reproducible-apk-tools.git
  - OUT=app/build/outputs/apk/release/app-release-unsigned.apk
  - reproducible-apk-tools/rm-files.py "$OUT" /outputs/unsigned.apk 'kotlin/*'
  - reproducible-apk-tools/inplace-fix.py --internal --zipalign fix-newlines /outputs/unsigned.apk 'META-INF/services/*'
```

The first line I had to add as we need to build an unsigned APK (once the build succeeds, the signature from your APK is copied over using `apksigcopier`, which should result in a byte-identical APK – thus confirming both APKs were built from exactly the same source, and nothing was added or removed with yours). The second line is clear. line 6 is needed as you build on Windows, and due to a bug in the tooling, some of your files have Windows line endings.

Now, the funny part is line 5:  somehow, our build contained additional Kotlin metadata which we had to remove. Here's the original diff before I added lines 5+6:

```diff
  -rw-r--r--  0.0 unx    10760 b-     4614 defN 1981-01-01 01:01:02 81bd88ae lib/x86_64/libandroidx.graphics.path.so
- -rw----     2.0 fat        6 b-        8 defN 1981-01-01 01:01:02 cae8a2ec META-INF/services/N2.w
- -rw----     2.0 fat        6 b-        8 defN 1981-01-01 01:01:02 c8ae1cb5 META-INF/services/O2.a
+ -rw----     2.0 fat        5 b-        7 defN 1981-01-01 01:01:02 c36e837e META-INF/services/N2.w
+ -rw----     2.0 fat        5 b-        7 defN 1981-01-01 01:01:02 e843d0bd META-INF/services/O2.a
+ -rw----     2.0 fat      865 b-      523 defN 1981-01-01 01:01:02 792b0911 kotlin/annotation/annotation.kotlin_builtins
+ -rw----     2.0 fat     3635 b-     1489 defN 1981-01-01 01:01:02 e686db83 kotlin/collections/collections.kotlin_builtins
+ -rw----     2.0 fat     2471 b-     1020 defN 1981-01-01 01:01:02 a60e181c kotlin/concurrent/atomics/atomics.kotlin_builtins
+ -rw----     2.0 fat      137 b-      116 defN 1981-01-01 01:01:02 51120a1b kotlin/coroutines/coroutines.kotlin_builtins
+ -rw----     2.0 fat      583 b-      355 defN 1981-01-01 01:01:02 e7264f53 kotlin/internal/internal.kotlin_builtins
+ -rw----     2.0 fat    18653 b-     5129 defN 1981-01-01 01:01:02 58e14695 kotlin/kotlin.kotlin_builtins
+ -rw----     2.0 fat     3349 b-     1256 defN 1981-01-01 01:01:02 a41ec3d8 kotlin/ranges/ranges.kotlin_builtins
+ -rw----     2.0 fat     2370 b-     1246 defN 1981-01-01 01:01:02 636244c3 kotlin/reflect/reflect.kotlin_builtins
  -rw----     0.0 fat    14912 b-     3432 defN 1981-01-01 01:01:02 730f1241 AndroidManifest.xml
```

I have no clue why our build has those files while yours does not. Well, it was easy to fix, so no action needed – though I'm still curious how this happens. But, now Dhizuku has the green shield up at IzzyOnDroid, showing it was confirmed as RB:

![Image](https://github.com/user-attachments/assets/4616c493-b48f-4177-bc4b-7639784b9e27)

(Amarok is shown along as these are the results of a search for "Dhizuku", which Amarok mentions in its app description as compatible). So: congrats! 🥳

**Comment by s1204IT** [comment]
Rebuilt fastlane/metadata and added it to Weblate.

All that's left is the icon and screenshots of the revised design on a smartphone and tablet.

**Comment by shuvashish76** [comment]
The full description is too short IMO. Closing #54

**Comment by s1204IT** [comment]
> The full description is too short IMO.

It's still in the proposal stage and has not been finalized.

**Comment by s1204IT** [comment]
#129 
Now that the icon is created, all that's left to do is fix the metadata, publish to the Play Store, and F-Droid will be on its way.

**Comment by IzzySoft** [comment]
Don't forget some screenshots in your fastlane. IIRC, PlayStore demands at least 2.

**Comment by s1204IT** [comment]
Yes, that's also included in the metadata I mentioned. There are seven of them available on the Play Store, but they're really just screenshots.

**Comment by s1204IT** [comment]
@IzzySoft Is there anything currently missing in fastlane metadata?

**Comment by IzzySoft** [comment]
Technically, all "mandatory" parts are there. Optionally, if you want to, you could add a featureGraphic, and some more things – but those are not _required_.

Comparing to what is set up here at IzzyOnDroid, the `full_description.txt` could be improved – but that may be partially subjective whether it's needed. Here's what IoD has:

![Image](https://github.com/user-attachments/assets/2c07dac2-daf4-4ba5-aaa0-a33699e6da47)

The current fulldesc is definitely clear to you, but for those not familiar with the topic, maybe a bit "cryptic". But your new screenshots are much better than what we had, so I went right into it and manually triggered a refresh here 🤗 (which updated the screenshots, and pulled in the icon plus the shortdescs for the other locales).

![Image](https://github.com/user-attachments/assets/4745e50a-8c54-4986-a87b-031257a4e174)

TL;DR: technically, all is fine. You might wish to make the fulldesc a little more "beginner friendly", though 😉

**Comment by s1204IT** [comment]
Okay, and thanks :)

Let's improve the full description.

**Comment by s1204IT** [comment]
This is just an example of text and I plan to add HTML tags if it is in good form.

**Comment by IzzySoft** [comment]
That would be great, yeah! But please also consider that people might not be willing to follow "yet another link" just to find out **what the app is about** – the link is of course fine for details on how it works and how to use it. Including a simple sentence with an easy example would help a lot.  From my screenshot above, and cleaned up a bit:

> An example use would be having a new Installer in the Aurora Store for allowing devices where normal package installation is blocked to install apps from third-party app stores.

And btw: mind to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link to  https://apt.izzysoft.de/packages/com.rosan.dhizuku from your Readme? We have "shields" as well, to fit next to your "Downloads" and "Latest" shields 😉

**Comment by s1204IT** [comment]
I see, so let me add a concrete example of how it can be used.

---

I had actually already been thinking about adding a label for IzzyOnDroid, but I am considering adding it at the same time as it is released on F-Droid and the Play Store.

**Comment by IzzySoft** [comment]
Making the badge-change in one go sounds reasonable, yes. Though Dhizuku is listed at IoD already for over a year 😉 Then let's hope the other two listings don't take too long 🙈

**Comment by theimpulson** [comment]
Linking this merge request here for consistency: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/22871

**Comment by s1204IT** [comment]
@theimpulson thanks!!

**Comment by s1204IT** [comment]
Hello!

We are ready to publish Dhizuku on the Play Store!

To do this, we need at least 12 testers to help us.

The way you can help us is to join the Google group below and join the testers.

Then install the test version of Dhizuku from the Play Store.

Join the group: 
https://groups.google.com/g/rosan-installer
Don't send any messages.

Join the test: https://play.google.com/apps/testing/com.rosan.dhizuku

Download:
https://play.google.com/store/apps/details?id=com.rosan.dhizuku

After you have installed it, please do not delete it for at least two weeks.

Thank you for your cooperation!

**Comment by IzzySoft** [comment]
Btw, @s1204IT @iamr0s 

```
com.rosan.dhizuku/metadata/bg/short_description.txt has 95 chars
com.rosan.dhizuku/metadata/el-GR/short_description.txt has 85 chars
com.rosan.dhizuku/metadata/pl-PL/short_description.txt has 90 chars
```

Fastlane specs limit those to 80 chars. Can you please take care to stay within those limits? Thanks in advance!
